src/Model/ContactRelation.php

```diff
@@ -0,0 +1,163 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Model;
+
+use Friendica\Core\Logger;
+use Friendica\Database\DBA;
+use Friendica\DI;
+use Friendica\Protocol\ActivityPub;
+use Friendica\Util\DateTimeFormat;
+use Friendica\Util\Strings;
+
+class ContactRelation
+{
+	/**
+	 * No discovery of followers/followings
+	 */
+	const DISCOVERY_NONE = 0;
+	/**
+	 * Discover followers/followings of local contacts
+	 */
+	const DISCOVERY_LOCAL = 1;
+	/**
+	 * Discover followers/followings of local contacts and contacts that visibly interacted on the system
+	 */
+	const DISCOVERY_INTERACTOR = 2;
+	/**
+	 * Discover followers/followings of all contacts
+	 */
+	const DISCOVERY_ALL = 3;
+
+	public static function store(int $target, int $actor, string $interaction_date)
+	{
+		if ($actor == $target) {
+			return;
+		}
+
+		DBA::update('contact-relation', ['last-interaction' => $interaction_date], ['cid' => $target, 'relation-cid' => $actor], true);
+	}
+
+	/**
+	 * Fetches the followers of a given profile and adds them
+	 *
+	 * @param string $url URL of a profile
+	 * @return void
+	 */
+	public static function discoverByUrl(string $url)
+	{
+		$contact_discovery = DI::config()->get('system', 'contact_discovery');
+
+		if ($contact_discovery == self::DISCOVERY_NONE) {
+			return;
+		}
+
+		$contact = Contact::getByURL($url);
+		if (empty($contact)) {
+			return;
+		}
+
+		if ($contact['last-discovery'] > DateTimeFormat::utc('now - 1 month')) {
+			Logger::info('Last discovery was less then a month before.', ['id' => $contact['id'], 'url' => $url, 'discovery' => $contact['last-discovery']]);
+			return;
+		}
+
+		if ($contact_discovery != self::DISCOVERY_ALL) {
+			$local = DBA::exists('contact', ["`nurl` = ? AND `uid` != ?", Strings::normaliseLink($url), 0]);
+			if (($contact_discovery == self::DISCOVERY_LOCAL) && !$local) {
+				Logger::info('No discovery - This contact is not followed/following locally.', ['id' => $contact['id'], 'url' => $url]);
+				return;
+			}
+
+			if ($contact_discovery == self::DISCOVERY_INTERACTOR) {
+				$interactor = DBA::exists('contact-relation', ["`relation-cid` = ? AND `last-interaction` > ?", $contact['id'], DBA::NULL_DATETIME]);
+				if (!$local && !$interactor) {
+					Logger::info('No discovery - This contact is not interacting locally.', ['id' => $contact['id'], 'url' => $url]);
+					return;
+				}
+			}
+		} elseif ($contact['created'] > DateTimeFormat::utc('now - 1 day')) {
+			Logger::info('Newly created contacs are not discovered to avoid DDoS attacks.', ['id' => $contact['id'], 'url' => $url, 'discovery' => $contact['created']]);
```

Comment:
```suggestion
			// Newly created contacts are not discovered to avoid DDoS attacks
			Logger::info('No discovery - Contact record is less than a day old.', ['id' => $contact['id'], 'url' => $url, 'discovery' => $contact['created']]);
```